Improve routes command to list all routes

adleong opened the pull request:
Fixes #1875 

This change improves the `linkerd routes` command in a number of important ways:

* The restriction on the type of the `--to` argument is lifted and any resource type can now be used.  Try `--to ns/books`, `--to po/webapp-ABCDEF`, `--to au/linkerd.io`, or even `--to svc`.
* All routes for the target will now be populated in the table, even if there are no Prometheus metrics for that route.
* [UNKNOWN] has been renamed to [DEFAULT]
* The `Service/Authority` column will now list `Service` in all cases except for when an authority target is explicitly requested.

```
$ linkerd routes deploy/traffic --to deploy/webapp
ROUTE                       SERVICE   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99
GET /                        webapp   100.00%   0.5rps          50ms         180ms         196ms
GET /authors/{id}            webapp   100.00%   0.5rps         100ms         900ms         980ms
GET /books/{id}              webapp   100.00%   0.9rps          38ms          93ms          99ms
POST /authors                webapp   100.00%   0.5rps          35ms          48ms          50ms
POST /authors/{id}/delete    webapp   100.00%   0.5rps          83ms         180ms         196ms
POST /authors/{id}/edit      webapp     0.00%   0.0rps           0ms           0ms           0ms
POST /books                  webapp    45.16%   2.1rps          75ms         425ms         485ms
POST /books/{id}/delete      webapp   100.00%   0.5rps          30ms          90ms          98ms
POST /books/{id}/edit        webapp    56.00%   0.8rps          92ms         875ms         975ms
[DEFAULT]                    webapp     0.00%   0.0rps           0ms           0ms           0ms
```

This is all made possible by a shift in the way we handle the destination resource.  When we get a request with a `ToResource`, we use the k8s API to find all Services which include at least one pod belonging to that resource.  We then fetch all service profiles for those services and display the routes from those serivce profiles.  

This shift in thinking also precipitates a change in the TopRoutes API where we no longer need special cases for `ToAll` (which can be specified by `--to au`) or `ToAuthority` (which can be specified by `--to au/<authority>`) and instead can use a `ToResource` to handle all cases.  We deprecate the old fields but continue to support them in the API for now.

grampelberg commented:
What am I doing wrong?

```
$ bin/go-run cli -n booksapp routes deploy/webapp
Error: No ServiceProfiles found for Services of deployment/webapp
$ kubectl -n linkerd get sp
NAME                                AGE
authors.default.svc.cluster.local   1m
books.default.svc.cluster.local     1m
webapp.default.svc.cluster.local    1m
```

grampelberg commented:
This is a *huge* improvement and I absolutely love it. 

Should the output be tweaked a little for things that have multiple returns? I'm seeing:

```bash
bin/go-run cli -n booksapp routes po
ROUTE                       SERVICE   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99
DELETE /authors/{id}.json   authors   100.00%   0.6rps          56ms          96ms          99ms
GET /authors.json           authors   100.00%   0.6rps           8ms          16ms          19ms
GET /authors/{id}.json      authors   100.00%   1.7rps           5ms           9ms          10ms
HEAD /authors/{id}.json     authors    52.02%   3.3rps           4ms           8ms          10ms
POST /authors.json          authors   100.00%   0.6rps          25ms          47ms          49ms
[DEFAULT]                   authors     0.00%   0.0rps           0ms           0ms           0ms
DELETE /books/{id}.json       books   100.00%   1.1rps          25ms          44ms          49ms
GET /books.json               books   100.00%   1.7rps           6ms          11ms          18ms
GET /books/{id}.json          books   100.00%   2.3rps           5ms           9ms          10ms
POST /books.json              books    52.71%   2.1rps          21ms          74ms          95ms
PUT /books/{id}.json          books    47.22%   1.2rps          75ms          98ms         100ms
[DEFAULT]                     books     0.00%   0.0rps           0ms           0ms           0ms
GET /                        webapp   100.00%   0.5rps          26ms          37ms          39ms
GET /authors/{id}            webapp   100.00%   0.5rps          22ms          29ms          30ms
GET /books/{id}              webapp   100.00%   1.2rps          21ms          32ms          38ms
POST /authors                webapp   100.00%   0.6rps          33ms          48ms          50ms
POST /authors/{id}/delete    webapp   100.00%   0.6rps          75ms          98ms         100ms
POST /authors/{id}/edit      webapp     0.00%   0.0rps           0ms           0ms           0ms
POST /books                  webapp    51.18%   2.1rps          29ms          76ms          95ms
POST /books/{id}/delete      webapp   100.00%   0.6rps          33ms          39ms          40ms
POST /books/{id}/edit        webapp    45.33%   1.2rps          75ms          98ms         100ms
[DEFAULT]                    webapp     0.00%   0.0rps           0ms           0ms           0ms
```

Could either add a new column for the resource name or do it the way tailing multiple files does (with a header).

adleong commented:
@rmars well, I guess it depends on if we care about maintaining backwards compatibility for a certain number of versions.  If we don't care at all, we can get rid of `GetToAll` et al.

adleong commented:
@grampelberg ^^ thoughts on backwards compatibility?  e.g. should it be possible to use a version of the CLI before this change with a version of the controller after this change?

grampelberg commented:
@adleong by "backwards compatibility" do you mean a future dashboard working with a past `web`?

grampelberg commented:
I don't have a problem with an old CLI returning no data. I'd rather that it didn't core dump though =)

adleong commented:
@grampelberg I've refactored things so that it shows a table per object and cleaned up a bunch of edge cases.  PTAL!

```
$ linkerd routes deploy --to deploy/authors
==> deployment/traffic <==
ROUTE                          SERVICE   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99
DELETE /authors/{id}.json      authors     0.00%   0.0rps           0ms           0ms           0ms
GET /authors.json              authors     0.00%   0.0rps           0ms           0ms           0ms
GET /authors/{id}.json         authors     0.00%   0.0rps           0ms           0ms           0ms
HEAD /authors/{id}.json        authors     0.00%   0.0rps           0ms           0ms           0ms
POST /authors.json             authors     0.00%   0.0rps           0ms           0ms           0ms
[DEFAULT]                      authors     0.00%   0.0rps           0ms           0ms           0ms
[DEFAULT]                   kubernetes     0.00%   0.0rps           0ms           0ms           0ms

==> deployment/webapp <==
ROUTE                          SERVICE   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99
DELETE /authors/{id}.json      authors   100.00%   0.4rps          50ms         170ms         194ms
GET /authors.json              authors   100.00%   0.4rps          10ms          38ms          40ms
GET /authors/{id}.json         authors   100.00%   1.3rps          10ms         315ms         383ms
HEAD /authors/{id}.json        authors     0.00%   0.0rps           0ms           0ms           0ms
POST /authors.json             authors   100.00%   0.4rps          15ms          29ms          30ms
[DEFAULT]                      authors     0.00%   0.0rps           0ms           0ms           0ms
[DEFAULT]                   kubernetes     0.00%   0.0rps           0ms           0ms           0ms

==> deployment/authors <==
ROUTE                          SERVICE   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99
DELETE /authors/{id}.json      authors     0.00%   0.0rps           0ms           0ms           0ms
GET /authors.json              authors     0.00%   0.0rps           0ms           0ms           0ms
GET /authors/{id}.json         authors     0.00%   0.0rps           0ms           0ms           0ms
HEAD /authors/{id}.json        authors     0.00%   0.0rps           0ms           0ms           0ms
POST /authors.json             authors     0.00%   0.0rps           0ms           0ms           0ms
[DEFAULT]                      authors     0.00%   0.0rps           0ms           0ms           0ms
[DEFAULT]                   kubernetes     0.00%   0.0rps           0ms           0ms           0ms

==> deployment/books <==
ROUTE                          SERVICE   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99
DELETE /authors/{id}.json      authors     0.00%   0.0rps           0ms           0ms           0ms
GET /authors.json              authors     0.00%   0.0rps           0ms           0ms           0ms
GET /authors/{id}.json         authors     0.00%   0.0rps           0ms           0ms           0ms
HEAD /authors/{id}.json        authors    53.90%   2.4rps           7ms          34ms          39ms
POST /authors.json             authors     0.00%   0.0rps           0ms           0ms           0ms
[DEFAULT]                      authors     0.00%   0.0rps           0ms           0ms           0ms
[DEFAULT]                   kubernetes     0.00%   0.0rps           0ms           0ms           0ms
```

grampelberg commented:
Oh I really like that!

adleong commented:
@siggy Great idea!

```
$ linkerd routes deploy --to svc/books
==> deployment/authors <==
ROUTE                     SERVICE   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99
DELETE /books/{id}.json     books   100.00%   0.4rps           9ms          27ms          30ms
GET /books.json             books   100.00%   0.4rps           7ms          10ms          10ms
GET /books/{id}.json        books     0.00%   0.0rps           0ms           0ms           0ms
POST /books.json            books     0.00%   0.0rps           0ms           0ms           0ms
PUT /books/{id}.json        books     0.00%   0.0rps           0ms           0ms           0ms
[DEFAULT]                   books     0.00%   0.0rps           0ms           0ms           0ms

==> deployment/webapp <==
ROUTE                     SERVICE   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99
DELETE /books/{id}.json     books   100.00%   0.4rps           8ms          10ms          10ms
GET /books.json             books   100.00%   0.9rps           8ms          18ms          20ms
GET /books/{id}.json        books   100.00%   1.7rps           7ms          17ms          19ms
POST /books.json            books    46.61%   2.0rps          18ms          43ms          49ms
PUT /books/{id}.json        books    59.57%   0.8rps          75ms          98ms         100ms
[DEFAULT]                   books     0.00%   0.0rps           0ms           0ms           0ms
```

siggy commented on controller/api/public/top_routes.go:
```diff
@@ -2,40 +2,207 @@ package public
+}
+
 func (s *grpcServer) TopRoutes(ctx context.Context, req *pb.TopRoutesRequest) (*pb.TopRoutesResponse, error) {
+	errRsp := validateRequest(req)
```
consider a `log.Debugf("TopRoutes request: %+v", req)` at the top of this method.

adleong commented on cli/cmd/routes.go:
```diff
@@ -100,39 +97,46 @@ func renderRouteStats(resp *pb.TopRoutesResponse, options *routesOptions) string
-			os.Exit(0)
+		for resource, table := range tables {
+			if len(tables) > 1 {
+				fmt.Fprintf(w, "==> %s <==\t\f", resource)
```
Tabwriter requires that all cells be terminated with a \t character.  The \f character is to terminate the column so that this header doesn't influence the ROUTE column beneath it

rmars commented on cli/cmd/routes.go:
```diff
@@ -100,39 +97,46 @@ func renderRouteStats(resp *pb.TopRoutesResponse, options *routesOptions) string
-		if len(table) == 0 {
-			fmt.Fprintln(os.Stderr, "No traffic found.  Does the service have a service profile?  You can create one with the `linkerd profile` command.")
-			os.Exit(0)
+		for resource, table := range tables {
```
TIOLI it would be really nice to sort the table keys here and map over the sorted resources; otherwise the order switches up with `watch` etc

siggy replied:
i agree, otherwise `linkerd routes` with a `watch` command isn't super useful (which i'm doing right now to test retries in this branch!)
